local echo (5/n): Create outbox messages on send

PIG208 opened the pull request:
This is stacked atop #1463, toward #1441.

No UI/user-facing change in this PR.

chrisbobbe reviewed (commented):
Thanks! Comments below.

PIG208 commented on lib/model/message.dart:
```diff
@@ -84,17 +268,124 @@ class MessageStoreImpl extends PerAccountStoreBase with MessageStore {
-    return _apiSendMessage(connection,
-      destination: destination,
+  Future<void> sendMessage({required MessageDestination destination, required String content}) async {
+    if (!debugOutboxEnable) {
```
Hmm, I think we still need to send the message, just without `queueId` and `localId`. Could you elaborate on how you would structure this?

PIG208 commented on lib/model/message.dart:
```diff
@@ -8,12 +12,161 @@ import 'message_list.dart';
+      assert(userIds.contains(selfUserId));
+    }
+    return switch (destination) {
+      StreamDestination() => StreamOutboxMessage(
+        localMessageId: localMessageId,
+        selfUserId: selfUserId,
+        conversation: StreamConversation(
+          destination.streamId,
+          destination.topic.interpretAsServer(
+            zulipFeatureLevel: zulipFeatureLevel,
+            realmEmptyTopicDisplayName: realmEmptyTopicDisplayName),
+          displayRecipient: null),
+        content: content,
+      ),
```
Yeah, good catch.

If we are to handle the updates, the initial setup probably wouldn't change. We could then have something along the lines of:

```dart
  void handleUpdateRealmSettingsEvent(UpdateRealmSettingsEvent event) {
    final affectedLocalMessageIds = <int>[];
    for (final outboxMessage in outboxMessages.values) {
      if (outboxMessage is! StreamOutboxMessage) continue;
      outboxMessage.conversation = outboxMessage.destination.topic.interpretAsServer(
        zulipFeatureLevel: zulipFeatureLevel,
        realmEmptyTopicDisplayName: realmEmptyTopicDisplayName);
      affectedLocalMessageIds.add(outboxMessage.localMessageId);
    }
    for (final view in _messageListViews) {
      view.handleOutboxMessageUpdate(affectedLocalMessageIds);
    }
  }
```

Just making `conversation` lazily evaluated from the current value of `PerAccountStore` wouldn't by itself be sufficient, since we still need to notify the message list views whenever such a change happens.

Right now, since we don't handle these events, I'm inclined to leave it to the issues you linked to.

gnprice replied:
I think having a comment noting this can happen is an adequate resolution. The effect is basically that the predicted ultimate topic can change between "(no topic)" and "general chat" and a different name for "general chat", right?

It should be pretty uncommon for that setting to change between when an attempted send happens and when the outbox-message goes away. There's also an unavoidable race that has the same effect: an admin could change the name of "general chat" (i.e. the value of `realmEmptyTopicDisplayName`) concurrently with the user making the send request, so that the setting in effect by the time the request arrives is different from the setting the client last heard about. So I think the comment doesn't need to be a TODO.

PIG208 commented on test/model/message_test.dart:
```diff
@@ -75,6 +93,365 @@ void main() {
+      // request was sucessful.
+      await store.handleEvent(eg.messageEvent(message,
+        localMessageId: outboxMessage.localMessageId));
+      check(store.outboxMessages).isEmpty();
+      check(outboxMessage)
+        ..state.equals(OutboxMessageLifecycle.sent)
+        ..hidden.isTrue();
```
The second sentence of the comment before `store.handleEvent` is meant to address this check:

```
      // Handle the event after the message is sent but before the debounce
      // timeout.  The outbox message should remain hidden since the send
      // request was successful.
```

Perhaps it will be clearer to move this right before the check.

PIG208 commented on lib/model/message.dart:
````diff
@@ -8,12 +12,161 @@ import 'message_list.dart';
+/// States of an [OutboxMessage] since its creation from a
+/// [MessageStore.sendMessage] call and before its eventual deletion.
+///
+/// ```
+///                              ┌─────────────────────────────────────┐
+///                              │                  Event received,    │
+///              Send            │                  or we abandoned    │
+///           immediately.       │      200.        the queue.         ▼
+/// (create) ──────────────► sending ────────► sent ──────────────► (delete)
+///                              │4xx,          │                      ▲
+///                              │other error,  │Reached       User    │
+///                              │or reached    │time limit.   cancels.│
+///                              │time limit.   ▼                      │
+///                              └───────────► failed ─────────────────┘
+/// ```
````
For the diagram update, started a discussion here: [#mobile > #F1441 Handle retry state machine @ 💬](https://chat.zulip.org/#narrow/channel/48-mobile/topic/.23F1441.20Handle.20retry.20state.20machine/near/2156104)

PIG208 commented:
Updated the PR! There are some pending questions mainly with regards to the state diagram and the send time limit. I introduced the new `waitPeriodExpired` state in this revision, with some changes to the states.

PIG208 commented:
Will be working on a new revision to reorganize some of the implementation code with further state machine changes.

PIG208 commented:
The PR has been updated to implement the state diagram discussed in chat: [#mobile > #F1441 Handle retry state machine @ 💬](https://chat.zulip.org/#narrow/channel/48-mobile/topic/.23F1441.20Handle.20retry.20state.20machine/near/2157168)

PIG208 commented:
(pushed another update to make timestamp used in sendMessage testable)

chrisbobbe commented on lib/model/message.dart:
```diff
@@ -1,19 +1,310 @@
+        content: content),
+    };
+
+    _outboxMessageDebounceTimers[localMessageId] = Timer(kLocalEchoDebounceDuration, () {
+      assert(outboxMessages.containsKey(localMessageId));
+      _outboxMessageDebounceTimers.remove(localMessageId);
+      _updateOutboxMessage(localMessageId, newState: OutboxMessageState.waiting);
+    });
+
+    _outboxMessageWaitPeriodTimers[localMessageId] = Timer(kSendMessageRetryWaitPeriod, () {
+      assert(outboxMessages.containsKey(localMessageId));
+      _outboxMessageWaitPeriodTimers.remove(localMessageId);
+      _updateOutboxMessage(localMessageId, newState: OutboxMessageState.waitPeriodExpired);
+    });
```
The `assert`s don't look right to me; wouldn't they throw if an outbox message was removed before the timer finished?, which can happen in this part of the state diagram:

```dart
///              Event received.
///              Or we abandoned the queue.
/// (any state) ────────────────────────────► (delete)
```

PIG208 replied:
`(delete)` doesn't have outgoing edges triggered by timeouts. So when the outbox message, for example, transition from `waiting` to `(delete)`, the timers should have been canceled too.

chrisbobbe commented on lib/model/message.dart:
```diff
@@ -1,19 +1,310 @@
+        queueId: queueId,
+        localId: localMessageId.toString());
+    } catch (e) {
+      // `localMessageId` is not necessarily in the store. This is because
+      // message event can still arrive before the send request fails to
+      // networking issues.
```
nit: maybe "fails with networking issues" or "fails because of networking issues"

PIG208 commented on lib/api/model/model.dart:
```diff
@@ -586,6 +595,30 @@ extension type const TopicName(String _value) {
   /// using [canonicalize].
   bool isSameAs(TopicName other) => canonicalize() == other.canonicalize();
 
+  /// Convert this topic to match how it would appear on a message object from
+  /// the server, assuming the topic is originally for a send-message request.
+  ///
+  /// For a client that does not support empty topics,
+  /// a modern server (FL>=334) would convert "(no topic)" and empty topics to
+  /// `store.realmEmptyTopicDisplayName`.
+  ///
+  /// See also: https://zulip.com/api/send-message#parameter-topic
+  TopicName interpretAsServer({
+    required int zulipFeatureLevel,
+    required String? realmEmptyTopicDisplayName,
+  }) {
+    if (zulipFeatureLevel < 334) {
+      assert(_value.isNotEmpty);
+      return this;
+    }
+    if (_value == kNoTopicTopic || _value.isEmpty) {
+      // TODO(#1250): this assumes that the 'support_empty_topics'
+      //   client_capability is false; update this when we set it to true
+      return TopicName(realmEmptyTopicDisplayName!);
+    }
+    return TopicName(_value);
+  }
```
I think the method itself will stay, since the server logic for converting literal `realmEmptyTopicDisplayName` into an empty string still applies when the client supports empty topics. I agree that we should have a `// TODO(server-10)` comment, for the feature level check.

Yeah, making sure that this is aware of whitespace-trimming is a good point.

PIG208 commented on lib/model/message.dart:
```diff
@@ -1,19 +1,310 @@
+        conversation: StreamConversation(
+          streamId,
+          topic.interpretAsServer(
+            // Because either of the values can get updated, the actual topic
+            // can change, for example, between "(no topic)" and "general chat",
+            // or between different names of "general chat".  This should be
+            // uncommon during the lifespan of an outbox message.
+            //
+            // There's also an unavoidable race that has the same effect:
+            // an admin could change the name of "general chat"
+            // (i.e. the value of realmEmptyTopicDisplayName) concurrently with
+            // the user making the send request, so that the setting in effect
+            // by the time the request arrives is different from the setting the
+            // client last heard about.  The realm update events do not have
+            // information about this race for us to update the prediction
+            // correctly.
+            zulipFeatureLevel: zulipFeatureLevel,
+            realmEmptyTopicDisplayName: realmEmptyTopicDisplayName),
```
The first paragraph is much clearer. Thanks!

(I should remind myself of updating this once #1250 changes how `interpretAsServer` behaves in terms of the topic predicted.)

PIG208 commented on lib/model/message.dart:
```diff
@@ -1,19 +1,310 @@
+    required OutboxMessageState newState,
+  }) {
+    final outboxMessage = outboxMessages[localMessageId];
+    if (outboxMessage == null || outboxMessage.state == newState) {
```
> Is there a legitimate need for callers to pass a new state that's the same as the old state? 

I think not. Removing this so that such an invalid state change can be caught with `set state` assertions.

chrisbobbe commented:
This is next on my queue; sorry I've let it sit for a bit!

chrisbobbe commented on lib/api/model/model.dart:
```diff
@@ -604,6 +613,33 @@ extension type const TopicName(String _value) {
+    required int zulipFeatureLevel,
+    required String? realmEmptyTopicDisplayName,
+  }) {
+    assert(_value.trim() == _value);
```
Good to have the assert, but let's mention the restriction on when this can be called. I guess the public-interface way of saying that is: it should only be called if the instance was constructed from a string without leading/trailing whitespace, right? (or "from a trimmed string" or similar)

chrisbobbe reviewed (commented):
Actually, could you send the first four commits as a new PR (just a small comment on those, below) and mark it for Greg's review?:

a19064924 api [nfc]: Add TopicName.interpretAsServer
bf8063a9c store [nfc]: Move zulip{FeatureLevel,Version} to PerAccountStoreBase
1f126181c test [nfc]: Generate timestamps
f8e1dab5e binding [nfc]: Add utcNow

That should help move things along faster as I'll need more time for the main commit:

cc0d77328 message: Create an outbox message on send; manage its states

chrisbobbe reviewed (commented):
Thanks, this is an exciting feature!!

Some general comments on the tests:

What's the reason for the `late OutboxMessage outboxMessage;`, which we check even after the outbox message is removed from the store? If we removed that variable and all the test code that depends on it, would we risk failing to catch a user-visible bug? I guess it's better if the `OutboxMessage` doesn't get changed after it's removed from the store…but it doesn't seem like it would cause any user-visible bug if it did, and it also seems pretty unlikely. (The app code shouldn't need to keep those objects around to compare them by identity, because they have unique IDs.)

Do the tests cover all possible state transitions and their causes? I think a helpful organization would be one test for each state transition, or more than one if the transition has multiple causes:

- hidden -> failed
- waiting -> failed
- waitPeriodExpired -> failed
- hidden -> waiting
- waiting -> waitPeriodExpired
- failed -> (delete) because draft was restored
- waitPeriodExpired -> (delete) because draft was restored
- hidden -> (delete) because event received
- waiting -> (delete) because event received
- failed -> (delete) because event received
- waitPeriodExpired -> (delete) because event received

I think this would make a really helpful starting point for when we work on the complete outbox system in the future, which will require adding and removing some transitions in the state diagram. It'll be easier to see how to update the tests if they're labeled with the current set of transitions.

gnprice commented on lib/model/message.dart:
```diff
@@ -73,15 +366,18 @@ class _EditMessageRequestStatus {
     // a use case for initializing MessageStore with nonempty [messages].
     : messages = {};
 
+  final String? realmEmptyTopicDisplayName;
```
(In particular the `TODO(#668)` is the thing that will help us be sure to spot this and bundle it up into that new realm-settings-etc substore when we're about to start updating realm settings, #668, so that it gets the benefit of those updates.)

PIG208 commented on lib/model/message.dart:
```diff
@@ -1,21 +1,309 @@
+        readBySender: true,
+        queueId: queueId,
+        localId: localMessageId.toString());
+    } catch (e) {
+      // `localMessageId` is not necessarily in the store. This is because the
+      // message event can still arrive, before the send request fails with
+      // networking issues.
+      _outboxMessageDebounceTimers.remove(localMessageId)?.cancel();
+      _outboxMessageWaitPeriodTimers.remove(localMessageId)?.cancel();
+      _updateOutboxMessage(localMessageId, newState: OutboxMessageState.failed);
+      rethrow;
+    }
```
Thanks. This is Neat!

Looking at `UpdateMachine`, I think this check will be quite granular.

I suppose a good structure will be adding this flag prior to the main commit, making some or all (?) `editMessage`, `handle…Event`, and friends `assert(!_disposed)` at the beginning, and return conditionally (`if (_disposed) return`) after async-gaps.

And then, I guess for private helpers like `_notifyMessageListViewsForOneMessage`, `assert(!_disposed)` at the start is not necessary.

chrisbobbe replied:
That sounds right to me.

PIG208 replied:
An interesting case is `unregisterMessageList`. Ultimately I think we want to have assert(!_disposed)` there, i.e. the message-list view should be unregistered before the store is disposed. Leaving a TODO similar to the one we have in `dispose`.

PIG208 commented on lib/model/message.dart:
```diff
@@ -1,21 +1,309 @@
+    }
+  }
+
+  void _handleMessageEventOutbox(MessageEvent event) {
+    if (event.localMessageId != null) {
+      final localMessageId = int.parse(event.localMessageId!, radix: 10);
+      // The outbox message can be missing if the user removes it (to be
+      // implemented in #1441) before the event arrives.
+      // Nothing to do in that case.
+      _outboxMessages.remove(localMessageId);
+      _outboxMessageDebounceTimers.remove(localMessageId)?.cancel();
+      _outboxMessageWaitPeriodTimers.remove(localMessageId)?.cancel();
+    }
+  }
```
Right, that will cause the message-list views to be noticed twice if there is an outbox message — once from  `removeOutboxMessage`, and another time from `view.handleMessageEvent`. That's perhaps fine.

chrisbobbe replied:
Oh I see; yeah, I forgot that `MessageStoreImpl.handleMessageEvent` already calls `view.handleMessageEvent` for all the `_messageListViews`.

I guess `removeOutboxMessage` isn't meant as an internal helper function, as I'd thought, right? Its purpose is just to handle these two state transitions, by responding to a user interaction:

- `failed` -> `(delete)`
- `waitPeriodExpired` -> `(delete)`

How about renaming `removeOutboxMessage` around that public-facing role? It's quite similar to `MessageStore.takeFailedEdit`, for edit-message, right:

```dart
  /// Forgets the failed edit request and returns the attempted new content.
  ///
  /// Should only be called when there is a failed request,
  /// per [getEditMessageErrorStatus].
  String takeFailedMessageEdit(int messageId);
```

except that it can also be called from a "wait-period expired" state, not only a "failed" state. Maybe `takeStaleOutboxMessage` or maybe just `takeOutboxMessage`? And have its dartdoc say when it's allowed to be called, i.e., just from the `failed` or `waitPeriodExpired` state.

PIG208 commented:
Thanks for the reviews! I have reorganized the tests and addressed the comments, after rebasing this on top of edit message changes :tada:

> What's the reason for the late OutboxMessage outboxMessage;, which we check even after the outbox message is removed from the store? If we removed that variable and all the test code that depends on it, would we risk failing to catch a user-visible bug? I guess it's better if the OutboxMessage doesn't get changed after it's removed from the store…but it doesn't seem like it would cause any user-visible bug if it did, and it also seems pretty unlikely. (The app code shouldn't need to keep those objects around to compare them by identity, because they have unique IDs.)

I agree that checking the state after the outbox message is removed from the store does not seem necessary. Removed those checks, and kept this late variable is mostly a helper for us to check the state of the outbox message.

> Do the tests cover all possible state transitions and their causes? I think a helpful organization would be one test for each state transition, or more than one if the transition has multiple causes

Yeah. It is not straightforward to confirm that these sets of tests cover all the state transitions. Your proposed way of organizing the tests does help a lot! I hope that this revision makes tests look shorter and more concise.

chrisbobbe commented on test/model/message_test.dart:
```diff
@@ -76,6 +94,303 @@ void main() {
+
+    late OutboxMessage outboxMessage;
+
+    Future<void> prepareSendMessageToSucceed({
+      MessageDestination? destination,
+      int? zulipFeatureLevel,
+    }) async {
+      message = eg.streamMessage(stream: stream);
+      await prepare(stream: stream, zulipFeatureLevel: zulipFeatureLevel);
+      await prepareMessages([eg.streamMessage(stream: stream)]);
+      connection.prepare(json: SendMessageResult(id: 1).toJson());
+      await store.sendMessage(
+        destination: destination ?? streamDestination, content: 'content');
+      outboxMessage = store.outboxMessages.values.single;
+      check(outboxMessage).state.equals(OutboxMessageState.hidden);
+    }
```
This differs from `prepareSendMessageToFail` in that it awaits the send-message request. Is that intentional? If so, would it be helpful to rename the function or add some dartdoc about that?

PIG208 replied:
Right, because this helper never needs to get a delay, unlike `prepareSendMessageToFail`. Perhaps we can name it to mention its immediate nature.

chrisbobbe commented on test/model/message_test.dart:
```diff
@@ -76,6 +94,303 @@ void main() {
+    }));
+
+    group('… -> failed', () {
+      test('hidden -> failed', () => awaitFakeAsync((async) async {
+        await prepareSendMessageToFail();
+        await check(failToSendMessageFuture).throws();
+        check(outboxMessage).state.equals(OutboxMessageState.failed);
+
+        // Wait till we reach at least [kSendMessageOfferRestoreWaitPeriod] after
+        // the send request was initiated.
+        async.elapse(kSendMessageOfferRestoreWaitPeriod);
+        async.flushTimers();
+        // The outbox message should stay in the failed state;
+        // it should not transition to waiting or waitPeriodExpired.
+        check(outboxMessage).state.equals(OutboxMessageState.failed);
+      }));
```
I see your reply at https://github.com/zulip/zulip-flutter/pull/1472#issuecomment-2892391723 :

> I agree that checking the state after the outbox message is removed from the store does not seem necessary. Removed those checks, and kept this late variable is mostly a helper for us to check the state of the outbox message.

It's actually unhelpful to use the `late outboxMessage` variable here, isn't it (?), because if there's a bug where the outbox message is removed from the store when it shouldn't be, this test won't catch it.

If all the uses of `outboxMessage` were replaced with lookups in the store (via a helper function if needed), I think that would be better, because we'd be testing the interface of this subsystem more directly.

PIG208 replied:
> because if there's a bug where the outbox message is removed from the store when it shouldn't be, this test won't catch it

Oh right. That makes sense. Yeah, that will make it more end-to-end.

chrisbobbe reviewed (commented):
Thanks, I think this is getting close! I like the new organization of the tests! Small comments below.

chrisbobbe commented on test/model/message_test.dart:
```diff
@@ -76,6 +94,302 @@ void main() {
+      check(store.outboxMessages).keys.length.equals(10);
+    });
+
+    void checkState(OutboxMessageState state) {
```
nit: how about `expected` or `expectedState` for the param's name, so the caller knows to pass a hard-coded value instead of some actual output from the code under test

chrisbobbe reviewed (commented):
Thanks, LGTM! One nit below, and I'll mark for Greg's review.

gnprice commented on lib/model/message.dart:
```diff
@@ -1,21 +1,332 @@
+
+/// Manages the outbox messages portion of [MessageStore].
+mixin _OutboxMessageStore on PerAccountStoreBase {
+  late final UnmodifiableMapView<int, OutboxMessage> outboxMessages =
+    UnmodifiableMapView(_outboxMessages);
+  final Map<int, OutboxMessage> _outboxMessages = {};
```
Neat! We should perhaps use this UnmodifiableMapView wrapper more.

gnprice commented on lib/model/message.dart:
```diff
@@ -1,21 +1,332 @@
+        oldState == OutboxMessageState.hidden,
+      OutboxMessageState.waitPeriodExpired =>
+        oldState == OutboxMessageState.waiting,
+      OutboxMessageState.failed =>
+        oldState == OutboxMessageState.hidden
+        || oldState == OutboxMessageState.waiting
+        || oldState == OutboxMessageState.waitPeriodExpired,
```
nit: equivalently,
```suggestion
      OutboxMessageState.failed => true,
```

right? Given the `newState != oldState` above.

Alternatively, I think with this condition in place here, that inequality check is redundant.

gnprice commented on lib/model/message.dart:
```diff
@@ -1,21 +1,332 @@
+
+  /// A fresh ID to use for [OutboxMessage.localMessageId],
+  /// unique within this instance.
+  int _nextLocalMessageId = 0;
```
Let's start at 1 rather than 0. That matches the practice in all the server's autoincrementing database indexes.

gnprice commented on lib/model/message.dart:
```diff
@@ -137,21 +478,29 @@ class MessageStoreImpl extends PerAccountStoreBase with MessageStore {
+    }
+    return _outboxSendMessage(
+      destination: destination, content: content,
+      realmEmptyTopicDisplayName: realmEmptyTopicDisplayName);
```
It looks like this doesn't conform to the constraint written on the `realmEmptyTopicDisplayName` getter, and enforced with an assert there. This will eagerly call that getter, before knowing whether the ZFL means its value will be needed.

I think the refactor that would be needed to make that assertion fully hold for this path is probably too much to be worth doing here directly. Instead, let's just bypass it (e.g., look at `_realmEmptyTopicDisplayName` directly) and leave a TODO comment.

I think probably the right solution is ultimately to have `TopicName.processLikeServer` (which is the method that consults this value and causes the need to pass it down here) to instead be a method on the per-account store, and more specifically on the RealmSettingStore (or similar) that we plan to set up in the future. Then it can consult that setting, as well as `zulipFeatureLevel`, directly off of that substore, and call the `realmEmptyTopicDisplayName` for itself only after it determines it's needed. But that substore is a post-launch refactor.

gnprice commented on lib/model/message.dart:
```diff
@@ -1,21 +1,332 @@
+        timestamp: now,
+        conversation: StreamConversation(
+          streamId,
+          topic.processLikeServer(
+            // Processing this just once on creating the outbox message
+            // allows an uncommon bug, because either of these values can change.
+            // During the outbox message's life, a topic processed from
```
Let's move this `topic.processLikeServer` call out to a little private helper method. That will keep the flow of the rest of this method's logic a bit easier to read.

gnprice commented on lib/model/message.dart:
```diff
@@ -1,21 +1,332 @@
+            realmEmptyTopicDisplayName: realmEmptyTopicDisplayName),
+          displayRecipient: null),
+        content: content),
+      DmDestination(:final userIds) => DmOutboxMessage(
```
How about giving `OutboxMessage` a factory constructor that takes an arbitrary `Conversation`, and dispatches between these two subclass constructors? That would allow this `switch` to get pushed down into computing just `conversation`, and the rest of these two constructor calls to get deduplicated. I think that'd make the structure of what this is doing a bit clearer.

The existing `OutboxMessage` constructor could get renamed to `OutboxMessage._` to make room for the nice factory constructor to have the plain `OutboxMessage` name.

gnprice commented on lib/model/message.dart:
```diff
@@ -1,21 +1,332 @@
+        content: content),
+    };
+
+    _outboxMessageDebounceTimers[localMessageId] = Timer(kLocalEchoDebounceDuration, () {
+      assert(!_disposed);
+      assert(outboxMessages.containsKey(localMessageId),
+        'The timer should have been canceled when the outbox message was removed.');
+      _outboxMessageDebounceTimers.remove(localMessageId);
+      _updateOutboxMessage(localMessageId, newState: OutboxMessageState.waiting);
+    });
```
Let's pull this callback out to its own method:
```suggestion
    _outboxMessageDebounceTimers[localMessageId] = Timer(
      kLocalEchoDebounceDuration, () => _handleOutboxDebounce(localMessageId));
```

That will help shorten this method and make its overall structure easier to see; and it'll also make clear there's a narrow interface between this overall method and the callback, with just the one bit of data `localMessageId` passing between them (vs. all the other parameters and locals that this method otherwise would offer to the callback's closure).

gnprice reviewed (commented):
Thanks @PIG208 for building this, and @chrisbobbe for the previous reviews! Exciting to see this approaching ready.

Generally this logic all looks good; various comments below. I haven't yet read the tests.

PIG208 commented on lib/model/message.dart:
```diff
@@ -1,21 +1,332 @@
+  /// [kLocalEchoDebounceDuration].
+  waiting,
+
+  /// The message was assumed not delivered after some time it was sent.
+  ///
+  /// This state can be reached when the message event hasn't arrived in
+  /// [kSendMessageOfferRestoreWaitPeriod] since the outbox message's creation.
+  waitPeriodExpired,
```
Yeah, this proposed behavior sounds better to me.

I think an easy implementation will just require us to cancel the timer that leads to `waitPeriodExpired` when the request is successful. At first I thought we might want to re-introduce that `sending` state from a previous draft, but I believe that shouldn't be necessary.

gnprice replied:
Sounds good!

PIG208 commented:
Thanks for the review! Updated the PR.

gnprice commented on lib/model/message.dart:
```diff
@@ -1,21 +1,332 @@
+}
+
+/// Manages the outbox messages portion of [MessageStore].
+mixin _OutboxMessageStore on PerAccountStoreBase {
```
bump — after MessageStoreImpl, not only MessageStore, and including all the new code so the MessageStore interface remains at the top :slightly_smiling_face:

(That way when the reader is thinking about aspects other than the outbox, for example the maintenance of the `messages` map itself, they can quickly go straight to those without skimming past the outbox logic; and conversely when they want to study how the outbox works, that's still all together in one place, namely the end of the file.)

gnprice commented on lib/model/message.dart:
```diff
@@ -137,21 +539,30 @@ class MessageStoreImpl extends PerAccountStoreBase with MessageStore {
+    }
+    return _outboxSendMessage(
+      destination: destination, content: content,
+      // TODO move [TopicName.processLikeServer] to a substore, eliminating this
+      realmEmptyTopicDisplayName: _realmEmptyTopicDisplayName);
```
nit: let's add a link to https://github.com/zulip/zulip-flutter/pull/1472#discussion_r2099069276 to expand on the details of this thought:
```suggestion
      // TODO move [TopicName.processLikeServer] to a substore, eliminating this;
      //   see https://github.com/zulip/zulip-flutter/pull/1472#discussion_r2099069276
      realmEmptyTopicDisplayName: _realmEmptyTopicDisplayName);
```

gnprice commented on lib/model/message.dart:
```diff
@@ -1,21 +1,180 @@
+  }) : _state = OutboxMessageState.hidden,
+       super(senderId: selfUserId);
+
+  // TODO(dart): This has to be a plain static method, because factories/constructors
+  //   do not support type parameters: https://github.com/dart-lang/language/issues/647
+  static OutboxMessage fromConversation(Conversation conversation, {
```
Ah right, indeed — so it can't quite have the short `OutboxMessage(…)` name like I suggested at https://github.com/zulip/zulip-flutter/pull/1472#discussion_r2099073411 above.

gnprice commented on lib/model/message.dart:
```diff
@@ -66,6 +234,219 @@ mixin MessageStore {
+        readBySender: true,
+        queueId: queueId,
+        localId: localMessageId.toString());
+      _outboxMessageWaitPeriodTimers.remove(localMessageId)?.cancel();
```
Cool, I guess this line is the implementation of https://github.com/zulip/zulip-flutter/pull/1472#discussion_r2099101759 .

gnprice commented on lib/model/message.dart:
```diff
@@ -66,6 +234,219 @@ mixin MessageStore {
+        readBySender: true,
+        queueId: queueId,
+        localId: localMessageId.toString());
+      _outboxMessageWaitPeriodTimers.remove(localMessageId)?.cancel();
```
This feels a bit mysterious on its own, so let's add a comment explaining why it's here:
```suggestion

      // The send request succeeded, so the message was definitely sent.
      // Cancel the timer that would have had us start presuming that the
      // send might have failed.
      _outboxMessageWaitPeriodTimers.remove(localMessageId)?.cancel();
```

gnprice replied:
and nit: move this outside of the try/catch, to make clear the try/catch is focused on exceptions from `_apiSendMessage` (where exceptions are expected) and not potential exceptions from this line (which shouldn't be possible). Discussion here:
https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#catch-specific

gnprice replied:
Writing out that comment also caused me to have the thought: what happens if the request succeeds and the state has *already* moved to `waitPeriodExpired`? The expiry means we've already started presuming that it probably failed, but suddenly we know for sure it actually succeeded.

If there isn't a quick change to make that would handle that, it's fine to leave a TODO comment for it instead.

PIG208 replied:
> and nit: move this outside of the try/catch, to make clear the try/catch is focused on exceptions from `_apiSendMessage` (where exceptions are expected) and not potential exceptions from this line (which shouldn't be possible). Discussion here: https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#catch-specific

Ah. This reminds me of adding a `if (_disposed)` check.

PIG208 replied:
> Writing out that comment also caused me to have the thought: what happens if the request succeeds and the state has _already_ moved to `waitPeriodExpired`? The expiry means we've already started presuming that it probably failed, but suddenly we know for sure it actually succeeded.
> 
> If there isn't a quick change to make that would handle that, it's fine to leave a TODO comment for it instead.


I think this would be a matter of implementing a new state transition:

 ```                                                                         
                             Got an [ApiRequestException].                   
          ┌──────┬────────────────────────────┬──────────► failed            
          │      │                            │              │               
          │      │      [sendMessage]         │              │               
 (create) │      │      request succeeds.     │              │               
    └► hidden   waiting ◄─────────────── waitPeriodExpired ──┴─────► (delete)
          │      ▲   │                     ▲               User restores     
          └──────┘   └─────────────────────┘               the draft.        
         Debounce     [sendMessage] request                                  
         timed out.   not finished when                                      
                      wait period timed out.                                 
                                                                             
              Event received.                                                
 (any state) ─────────────────► (delete)                                     
 ```                                                                         

I guess we can't really convey that `waitPeriodExpired` cannot be reached again without having a variant of `waiting`, but we can imply that from the dart doc of `waitPeriodExpired`:

```dart
  /// The [sendMessage] HTTP request did not finish in time and the user is
  /// invited to retry it.
  ///
  /// This state can be reached when the request has not finished
  /// [kSendMessageOfferRestoreWaitPeriod] since the outbox message's creation.
  waitPeriodExpired,
```

PIG208 replied:
Added a new commit `message: Avoid double-sends after send-message request succeeds`  addressing this. I found it to be a good way seeing how we can incrementally add features to this state machine.

gnprice reviewed (commented):
Thanks for the revision! Comments below. I'll also go take a quick initial look at the tests.

gnprice commented on test/model/message_test.dart:
```diff
@@ -76,6 +94,314 @@ void main() {
+    test('smoke DM: hidden -> waiting -> (delete)', () => awaitFakeAsync((async) async {
+      await prepareOutboxMessage(destination: DmDestination(
+        userIds: [eg.selfUser.userId, eg.otherUser.userId]));
+      checkState(expectedState: OutboxMessageState.hidden);
```
nit: can put these in more canonical `checks` style:
```suggestion
      checkState().equals(OutboxMessageState.hidden);
```

gnprice commented on test/model/message_test.dart:
```diff
@@ -76,6 +94,314 @@ void main() {
+      check(store.outboxMessages).isEmpty();
+    }));
+
+    test('smoke stream message: hidden -> waiting -> (delete)', () => awaitFakeAsync((async) async {
+      await prepareOutboxMessage();
+      checkState(expectedState: OutboxMessageState.hidden);
+
+      async.elapse(kLocalEchoDebounceDuration);
+      checkState(expectedState: OutboxMessageState.waiting);
```
I like this structure where the helpers center on a single outbox-message. It helps keep the tests focused, since the bulk of them are all about a single outbox-message (because there's very little between-message interaction to be tested).

gnprice reviewed (commented):
Cool, and generally the tests look good.

I've read the miscellaneous test changes outside the main block in `message_test.dart`, and skimmed over the latter; just a couple of comments below. I'll leave for another day reading through that main block of tests in full.

gnprice commented on test/model/message_test.dart:
```diff
@@ -76,6 +94,341 @@ void main() {
+      check(store.outboxMessages).isEmpty();
+    }));
+
+    test('smoke stream message: hidden -> waiting -> (delete)', () => awaitFakeAsync((async) async {
+      await prepareOutboxMessage();
```
nit: since the test description specifically calls for a stream message, the test should explicitly make it a stream message

(the bulk of the other tests below can cheerfully go on using the default that `prepareOutboxMessage` supplies, since they don't care about the specifics of the message)

gnprice commented on test/model/message_test.dart:
```diff
@@ -76,6 +94,341 @@ void main() {
+      async.elapse(kLocalEchoDebounceDuration);
+      checkState().equals(OutboxMessageState.waiting);
+
+      await store.handleEvent(eg.messageEvent(
+        eg.dmMessage(from: eg.selfUser, to: [eg.otherUser]),
+        localMessageId: store.outboxMessages.keys.single));
```
nit: could tighten these tests further (going in the same direction noted at https://github.com/zulip/zulip-flutter/pull/1472#discussion_r2099318648) with a helper for this step:
```suggestion
      await receiveMessage(eg.dmMessage(from: eg.selfUser, to: [eg.otherUser]));
```

Could even have that helper default to the shared `message` local, since that's what most of these tests use.

gnprice commented on test/model/message_test.dart:
```diff
@@ -76,6 +94,341 @@ void main() {
+          localMessageId: store.outboxMessages.keys.single));
+        check(store.outboxMessages).isEmpty();
+
+        // Complete the send request.  There should be no error despite
+        // the send request failure, because the outbox message is not
+        // in the store any more.
+        await check(outboxMessageFailFuture).completes();
```
Cool, this is an important subtle case to test.

(Important because it could be fairly common if the user is on a flaky connection — if the rate of requests randomly failing is anywhere between like 20% and 80%, then a good fraction of the time, the send will make it through, the response won't and so the request will eventually throw when the connection times out, and in the meantime the event will make it through. Important because also it'd be pretty annoying to get a useless error dialog when nothing actually went wrong.)

PIG208 replied:
Yeah, exactly. This is for covering the change from Chris here: https://github.com/zulip/zulip-flutter/pull/1472#discussion_r2093703191

gnprice commented on test/model/message_test.dart:
```diff
@@ -76,6 +94,341 @@ void main() {
+      message = eg.streamMessage(stream: stream);
+      await prepare(stream: stream);
+      await prepareMessages([eg.streamMessage(stream: stream)]);
+      connection.prepare(apiException: eg.apiBadRequest(), delay: delay);
```
nit: for better realism, make it an error that's consistent with the request having actually succeeded on the server side:
```suggestion
      connection.prepare(httpException: SocketException('failed'), delay: delay);
```

That way it's consistent with the test cases where a corresponding MessageEvent arrives even though the request failed. That also may help a future reader in understanding why that scenario can even happen (which is important to understand since we have some logic carefully written to specifically handle it).

gnprice commented on test/model/message_test.dart:
```diff
@@ -76,6 +94,341 @@ void main() {
+        destination: StreamDestination(stream.streamId, TopicName('(no topic)')),
+        zulipFeatureLevel: 370);
+      async.elapse(kLocalEchoDebounceDuration);
+      check(store.outboxMessages.values.single)
```
nit: use checks-getters
```suggestion
      check(store.outboxMessages).values.single
```

gnprice commented on lib/model/message.dart:
```diff
@@ -435,4 +488,414 @@ class MessageStoreImpl extends PerAccountStoreBase with MessageStore {
+    // Cancel the timer that would have had us start presuming that the
+    // send might have failed.
+    _outboxMessageWaitPeriodTimers.remove(localMessageId)?.cancel();
+    if (_outboxMessages[localMessageId]!.state
+          == OutboxMessageState.waitPeriodExpired) {
```
Neat, glad that change turned out to be clean to make! (Following up on https://github.com/zulip/zulip-flutter/pull/1472#discussion_r2101148783.)

The diffstat is:
 lib/model/message.dart       | 28 ++++++++++++++++++++--------
 test/model/message_test.dart | 28 ++++++++++++++++++++++++++++
 2 files changed, 48 insertions(+), 8 deletions(-)

and outside the new test, most of that is comments; I count 6+/1- other than tests and comments.

gnprice reviewed (commented):
Cool, and here's a full review of the remainder: the tests, and the new commit at the end. Generally all looks good.

PIG208 commented on test/model/message_test.dart:
```diff
@@ -76,6 +94,341 @@ void main() {
+      checkState().equals(OutboxMessageState.waiting);
+    }));
+
+    test('waitPeriodExpired -> waiting and never return to waitPeriodExpired', () => awaitFakeAsync((async) async {
```
There are two cases of "waiting -> waitPeriodExpired -> (delete) …" that cover this. But I think including the first edge will still be helpful because it shows that there is a loop.

gnprice replied:
Cool. Yeah, I see those as focused on the later arrow that goes to "(delete)"; partly because they're tucked inside the group "… -> (delete)". When looking for the test that covers this arrow of the state diagram, this seems like the more natural spot to find it.

(Ideally in fact I think there'd be a case here that was explicitly focused on just that arrow, without a subsequent arrow.)

gnprice commented on test/model/message_test.dart:
```diff
@@ -76,6 +94,335 @@ void main() {
+
+    final localMessageIds = store.outboxMessages.keys.toList();
+    store.takeOutboxMessage(localMessageIds.removeAt(5));
+    check(store.outboxMessages.keys).deepEquals(localMessageIds);
```
nit: use checks-getter (like https://github.com/zulip/zulip-flutter/pull/1472#discussion_r2101317278 )
```suggestion
    check(store.outboxMessages).keys.deepEquals(localMessageIds);
```

gnprice reviewed (commented):
Thanks for the revision! Just one nit remaining.

PIG208 commented:
In addition to the nit, added a test that targets "waiting -> waitPeriodExpired" specifically, for https://github.com/zulip/zulip-flutter/pull/1472#discussion_r2106517535, dropping the emphasis on "waiting -> waitPeriodExpired" from those two "waiting -> waitingPeriodExpired -> (delete)" tests.

gnprice commented:
Thanks! Those changes all look good; merging.
